pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	recoveryPod, ok := recoveryPodObj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("unsupported type: internal recovery pod is not type *corev1.Pod but %T", recoveryPod)
+	}
+
+	return recoveryPod, nil
+}
+
+func (s *Apiserver) Create() error {
+	resourceDirPath, err := s.GetResourceDirPath()
+	if err != nil {
+		return fmt.Errorf("failed to get resource dir: %v", err)
+	}
+
+	recoveryDirPath, err := s.GetRecoveryDirPath()
+	if err != nil {
+		return fmt.Errorf("failed to get recovery dir: %v", err)
+	}
+
+	// Copy certs for accessing etcd
+	for src, dest := range map[string]string{
+		"secrets/etcd-client/tls.key":              "etcd-client.key",
+		"secrets/etcd-client/tls.crt":              "etcd-client.crt",
+		"configmaps/etcd-serving-ca/ca-bundle.crt": "etcd-serving-ca-bundle.crt",
+	} {
+		err = copyFile(path.Join(resourceDirPath, src), path.Join(recoveryDirPath, dest))
+		if err != nil {
+			return err
+		}
+	}
+
+	// We are creating only temporary certificates to start the recovery apiserver.
+	// A week seem reasonably high for a debug session, while it is easy to create a new one.
+	certValidity := 7 * 24 * time.Hour
+
+	// Create root CA
+	rootCaConfig, err := crypto.MakeSelfSignedCAConfigForDuration("localhost", certValidity)
+	if err != nil {
+		return fmt.Errorf("failed to create root-signer CA: %v", err)
+	}
+
+	servingCaCertPath := path.Join(recoveryDirPath, "serving-ca.crt")
+	err = rootCaConfig.WriteCertConfigFile(servingCaCertPath, path.Join(recoveryDirPath, "serving-ca.key"))
+	if err != nil {
+		return fmt.Errorf("failed to write root-signer files: %v", err)
+	}
+
+	// Create config for recovery apiserver
+	recoveryConfigBytes, err := v311_00_assets.Asset(RecoveryConfigAsset)
+	if err != nil {
+		return fmt.Errorf("fail to find internal recovery config asset %q: %v", RecoveryConfigAsset, err)
+	}
+
+	recoveryConfigPath := path.Join(recoveryDirPath, RecoveryCofigFileName)
+	err = ioutil.WriteFile(recoveryConfigPath, recoveryConfigBytes, 644)
+	if err != nil {
+		return fmt.Errorf("failed to write recovery config %q: %v", recoveryConfigPath, err)
+	}
+
+	recoveryPod, err := s.recoveryPod()
+	if err != nil {
+		return fmt.Errorf("failed to create recovery pod: %v", err)
+	}
+
+	recoveryPodBytes, err := yaml.Marshal(recoveryPod)
+	if err != nil {
+		return fmt.Errorf("failed to marshal recovery pod: %v", err)
+	}
+
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err = ioutil.WriteFile(recoveryPodManifestPath, recoveryPodBytes, 644)
+	if err != nil {
+		return fmt.Errorf("failed to write recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// Create client cert
+	ca := crypto.CA{
+		Config:          rootCaConfig,
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+	}
+
+	// Create client certificates for system:admin
+	clientCert, err := ca.MakeClientCertificateForDuration(&user.DefaultInfo{Name: "system:admin"}, certValidity)
```

Comment:
// we are lazy and reuse the serving CA as client CA

Reply:
> // we are lazy and reuse the serving CA as client CA

I really hate that, but I won't fight it here for now.